engine/src/test/java/io/zeebe/engine/state/WorkflowExecutionCleanZeebeStateTest.java

```diff
@@ -0,0 +1,618 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import io.zeebe.engine.processing.message.MessageObserver;
+import io.zeebe.engine.state.instance.ElementInstanceState;
+import io.zeebe.engine.state.instance.EventScopeInstanceState;
+import io.zeebe.engine.state.instance.IncidentState;
+import io.zeebe.engine.state.instance.JobState;
+import io.zeebe.engine.state.instance.TimerInstanceState;
+import io.zeebe.engine.state.instance.VariablesState;
+import io.zeebe.engine.state.message.MessageStartEventSubscriptionState;
+import io.zeebe.engine.state.message.MessageState;
+import io.zeebe.engine.state.message.MessageSubscriptionState;
+import io.zeebe.engine.state.message.WorkflowInstanceSubscriptionState;
+import io.zeebe.engine.util.EngineRule;
+import io.zeebe.model.bpmn.Bpmn;
+import io.zeebe.protocol.record.intent.IncidentIntent;
+import io.zeebe.protocol.record.intent.JobIntent;
+import io.zeebe.protocol.record.intent.MessageStartEventSubscriptionIntent;
+import io.zeebe.protocol.record.intent.WorkflowInstanceIntent;
+import io.zeebe.protocol.record.value.BpmnElementType;
+import io.zeebe.test.util.record.RecordingExporter;
+import java.time.Duration;
+import java.util.List;
+import java.util.Map;
+import org.awaitility.Awaitility;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+
+public final class WorkflowExecutionCleanZeebeStateTest {
+
+  private static final String PROCESS_ID = "workflow";
+
+  @Rule public EngineRule engineRule = EngineRule.singlePartition();
+
+  private ElementInstanceState elementInstanceState;
+  private TimerInstanceState timerState;
+  private EventScopeInstanceState eventScopeInstanceState;
+  private JobState jobState;
+  private IncidentState incidentState;
+  private MessageState messageState;
+  private MessageSubscriptionState messageSubscriptionState;
+  private WorkflowInstanceSubscriptionState workflowInstanceSubscriptionState;
+  private MessageStartEventSubscriptionState messageStartEventSubscriptionState;
+  private VariablesState variablesState;
+
+  @Before
+  public void init() {
+    final var zeebeState = engineRule.getZeebeState();
+
+    final var workflowState = zeebeState.getWorkflowState();
+    elementInstanceState = workflowState.getElementInstanceState();
+    variablesState = elementInstanceState.getVariablesState();
+    timerState = workflowState.getTimerState();
+    eventScopeInstanceState = workflowState.getEventScopeInstanceState();
+
+    jobState = zeebeState.getJobState();
+    incidentState = zeebeState.getIncidentState();
+    messageState = zeebeState.getMessageState();
+    messageSubscriptionState = zeebeState.getMessageSubscriptionState();
+    workflowInstanceSubscriptionState = zeebeState.getWorkflowInstanceSubscriptionState();
+    messageStartEventSubscriptionState = zeebeState.getMessageStartEventSubscriptionState();
+
+    assertThatStateIsEmpty();
+  }
+
+  @Test
+  public void testWorkflowWithServiceTask() {
```

Comment:
Can we have one test with input mapping? Does it make a difference?

Reply:
> Can we have one test with input mapping? Does it make a difference?

There is one test case with input and output mappings. It should cover the cases.

> I would suggest that we change it to something like shouldEndWithCleanStateOnWorkflowWithServiceTask

Usually, I would agree. But in this case, all tests verify that the state is empty at the end - just with different parameters. It could be a `ParameterizedTest` but the workflow and steps are too different. 

If we're using `shouldEndWithCleanStateOnWorkflowWithServiceTask` then the major part `shouldEndWithCleanStateOnWorkflowWith` is equal to all tests. So, I would like to shrink the name to what is different. 

Do you agree with it?

Reply:
What I normally then do is to name the Test as the common thing and as you meant use the difference as test method name, but still with `should` prefix. 

For example `AlwaysCleanStateTest` `#shouldBeTrueWithWorkflowServiceTask`, `#shouldBeTrueWithTimerEvent` etc.

Reply:
I did not find a better way for the naming. Also, a parameterized test doesn't work nicely here.

Reply:
ok thanks for trying it